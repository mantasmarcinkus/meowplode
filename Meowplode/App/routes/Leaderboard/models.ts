export interface Player {
  name: string;
  dateCreated: string;
  result: boolean;
  successfullyAnswered: number;
}
