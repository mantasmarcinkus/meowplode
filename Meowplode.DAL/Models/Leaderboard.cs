using System;

namespace Meowplode.DAL.Models
{
    public class Leaderboard : BaseModel
    {
        public string Name { get; set; }

        public DateTime DateCreated { get; set; }

        public bool Result { get; set; }

        public int SuccessfullyAnswered { get; set; }
    }
}
