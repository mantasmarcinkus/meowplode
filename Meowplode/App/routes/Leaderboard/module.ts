import { Action, Dispatch } from "redux";
import { Player } from "./models";
export const LEADERBOARD_GET = "LEADERBOARD_GET";

export function getLeaderboard() {
  return (dispatch: Dispatch<any>) => {
    fetch(`api/leaderboard`)
      .then(response => response.json())
      .then(players => {
        dispatch({
          payload: players,
          type: LEADERBOARD_GET,
        });
      });
  };
}
export interface ActionCreators {
  getLeaderboard(): (dispatch: Dispatch<any>) => void;
}
const ACTION_HANDLERS: { [key: string]: any } = {
  [LEADERBOARD_GET]: (state: State, action: Action & { payload: Player[] }) => {
    return {
      ...state,
      players: action.payload,
    };
  },
};

export interface State {
  players: Player[] | null;
}

const initialState: State = {
  players: null,
};

export default function reducer(state = initialState, action: Action) {
  const handler = ACTION_HANDLERS[action.type];

  return handler ? handler(state, action) : state;
}
