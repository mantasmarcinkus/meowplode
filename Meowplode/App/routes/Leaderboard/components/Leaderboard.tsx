import * as React from "react";
import { Dispatch } from "redux";
import { Player } from "../models";
import { ActionCreators, State } from "../module";

const Cat = require("../assets/cat.png");
const Tombstone = require("../assets/tombstone.png");

const QUESTION_COUNT = 10;

export interface Props extends Partial<ActionCreators & State> {
  name: string;
}
export const Leaderboard = (props: Props) => {
  if (!props.players) {
    return <div>Loading...</div>;
  } else if (props.players.length === 0) {
    return (null);
  }

  const listOfPlayers = props.players.map(x => (
    <div key={x.dateCreated} className="leaderboard-row">
      <div className="leaderboard-image">
        <img alt="result" src={x.result ? Cat : Tombstone} height="24" />
      </div>
      <div className="leaderboard-name">
        <span style={{ fontWeight: props.name === x.name ? "bold" : "normal" }}>{x.name}</span>
        <span> ({x.successfullyAnswered}/{QUESTION_COUNT})</span>
      </div>
    </div>));
  return (
    <div>
      <div style={{ fontSize: 18, marginBottom: 7 }}>How did other players do?</div>
      <div className="leaderboard">
        {listOfPlayers}
      </div>
    </div>
  );
};

export default Leaderboard;
