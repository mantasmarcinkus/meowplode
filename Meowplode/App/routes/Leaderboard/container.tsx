import * as React from "react";
import { connect } from "react-redux";
import Leaderboard, { Props } from "./components/Leaderboard";
import { getLeaderboard, State } from "./module";

class Container extends React.Component<Props, {}> {
  public render() { return <Leaderboard {...this.props} />; }
  public componentWillMount() {
    if (this.props.getLeaderboard) {
      this.props.getLeaderboard();
    }
  }
}

const mapStateToProps = ({ leaderboard: { players } }: { leaderboard: State }, { name }: { name: string }) => {
  return {
    name,
    players,
  };
};

const mapDispatchToProps = {
  getLeaderboard,
};

export default connect(mapStateToProps, mapDispatchToProps)(Container as any);
