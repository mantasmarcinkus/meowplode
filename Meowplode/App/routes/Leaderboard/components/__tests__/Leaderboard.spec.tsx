import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import Leaderboard from "../Leaderboard";

jest.mock("../ImageLoader");

const testProps = {
  name: "text",
  players: [
    { dateCreated: "date1", result: true, successfullyAnswered: 10, name: "text" },
    { dateCreated: "date2", result: false, successfullyAnswered: 7, name: "text2" },
  ],
};
describe("<Leaderboard />", () => {
  let wrapper: ShallowWrapper<any, any>;
  beforeEach(() => {
    wrapper = shallow(<Leaderboard name="text"/>);
  });
  it("renders with name only", () => {
    expect(wrapper.render()).toMatchSnapshot();
  });
  it("renders with name and players", () => {
    expect(wrapper.render()).toMatchSnapshot();
  });
});
