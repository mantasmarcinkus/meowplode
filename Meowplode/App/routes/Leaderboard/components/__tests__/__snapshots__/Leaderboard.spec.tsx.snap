// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Leaderboard /> renders with name and players 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "Loading...",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;

exports[`<Leaderboard /> renders with name only 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "Loading...",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;
